doc/Data Plane Code Generation/ServiceDrivenEvolution.md

```diff
@@ -0,0 +1,177 @@
+# Service driven evolution
+
+As a service evolves with non-breaking changes in the REST API, it may or may not cause breaking changes in the generated SDK. For instance, adding a new method in an existing path or in a new path won't cause breaking changes in the new SDK. But some changes like adding an optional query parameter could cause breaking changes in the generated SDK. In order to make the SDK backward compatible, we can add customized code such as overload methods. The following sections will show some examples of such scenarios and how to handle them.
+
+## A method gets a new optional parameter
+
+Generated code in a V1 client with a required query parameter and an optional query parameter:
+
+``` C#
+public virtual async Task<Response> PutRequiredOptionalAsync(string requiredParam, string optionalParam = null, RequestContext context = null)
+{
+    Argument.AssertNotNull(requiredParam, nameof(requiredParam));
+
+    using var scope = ClientDiagnostics.CreateScope("ParamsClient.PutRequiredOptional");
+    scope.Start();
+    try
+    {
+        using HttpMessage message = CreatePutRequiredOptionalRequest(requiredParam, optionalParam, context);
+        return await _pipeline.ProcessMessageAsync(message, context).ConfigureAwait(false);
+    }
+    catch (Exception e)
+    {
+        scope.Failed(e);
+        throw;
+    }
+}
+```
+
+Generated code in a V2 client with a new optional query parameter:
+
+``` C#
+public virtual async Task<Response> PutRequiredOptionalAsync(string requiredParam, string optionalParam = null, string newParameter = null, RequestContext context = null)
+{
+    Argument.AssertNotNull(requiredParam, nameof(requiredParam));
+
+    using var scope = ClientDiagnostics.CreateScope("ParamsClient.PutRequiredOptional");
+    scope.Start();
+    try
+    {
+        using HttpMessage message = CreatePutRequiredOptionalRequest(requiredParam, optionalParam, newParameter, context);
+        return await _pipeline.ProcessMessageAsync(message, context).ConfigureAwait(false);
+    }
+    catch (Exception e)
+    {
+        scope.Failed(e);
+        throw;
+    }
+}
+```
+
+Since `context` is always the last parameter, V2 inserts the new optional parameter before `context` parameter and this could break code that works with V1 such as `client.PutRequiredOptionalAsync("requiredParam", "optionalParam", ErrorOptions.NoThrow)` (`ErrorOptions.NoThrow` was passed in as `context` parameter but now it's treated as `newParameter`). We can manually add an overload method in V2 which transforms the optional parameters in V1 to required parameters to make V2 backward compatible with V1:
+
+``` C#
+public virtual async Task<Response> PutRequiredOptionalAsync(string requiredParam, string optionalParam, RequestContext context)
+{
+    Argument.AssertNotNull(requiredParam, nameof(requiredParam));
```

Comment:
Do we need to validate other parameters of this function?

Reply:
https://azure.github.io/azure-sdk/general_implementation.html#parameter-validation